<a href="https://colab.research.google.com/github/geligon/geligon/blob/main/AStarMetData_GL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Create Environment ⚓

In [1]:
# Import libraries
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import missingno as msno
import datetime as dt


Read In Data ♥

In [2]:
df=pd.read_csv("MetDataEX.csv")


Overview of Data 👁

In [3]:
df.head()

Object Number  Is Highlight  Is Timeline Work  Is Public Domain  Object ID  \
0    1979.486.1         False             False             False          1   
1    1980.264.5         False             False             False          2   
2      67.265.9         False             False             False          3   
3     67.265.10         False             False             False          4   
4     67.265.11         False             False             False          5   

  Gallery Number         Department  AccessionYear Object Name  \
0            NaN  The American Wing         1979.0        Coin   
1            NaN  The American Wing         1980.0        Coin   
2            NaN  The American Wing         1967.0        Coin   
3            NaN  The American Wing         1967.0        Coin   
4            NaN  The American Wing         1967.0        Coin   

                          Title  ... River  Classification  \
0  One-dollar Liberty Head Coin  ...   NaN             NaN   
1  Ten-dollar Liberty Head Coin  ...   NaN             NaN   
2    Two-and-a-Half Dollar Coin  ...   NaN             NaN   
3    Two-and-a-Half Dollar Coin  ...   NaN             NaN   
4    Two-and-a-Half Dollar Coin  ...   NaN             NaN   

   Rights and Reproduction                                     Link Resource  \
0                      NaN  http://www.metmuseum.org/art/collection/search/1   
1                      NaN  http://www.metmuseum.org/art/collection/search/2   
2                      NaN  http://www.metmuseum.org/art/collection/search/3   
3                      NaN  http://www.metmuseum.org/art/collection/search/4   
4                      NaN  http://www.metmuseum.org/art/collection/search/5   

   Object Wikidata URL  Metadata Date  \
0                  NaN            NaN   
1                  NaN            NaN   
2                  NaN            NaN   
3                  NaN            NaN   
4                  NaN            NaN   

                                 Repository Tags Tags AAT URL  \
0  Metropolitan Museum of Art, New York, NY  NaN          NaN   
1  Metropolitan Museum of Art, New York, NY  NaN          NaN   
2  Metropolitan Museum of Art, New York, NY  NaN          NaN   
3  Metropolitan Museum of Art, New York, NY  NaN          NaN   
4  Metropolitan Museum of Art, New York, NY  NaN          NaN   

  Tags Wikidata URL  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4               NaN  

[5 rows x 54 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12549 entries, 0 to 12548
Data columns (total 54 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Object Number            12549 non-null  object 
 1   Is Highlight             12549 non-null  bool   
 2   Is Timeline Work         12549 non-null  bool   
 3   Is Public Domain         12549 non-null  bool   
 4   Object ID                12549 non-null  int64  
 5   Gallery Number           8559 non-null   object 
 6   Department               12549 non-null  object 
 7   AccessionYear            12283 non-null  float64
 8   Object Name              12549 non-null  object 
 9   Title                    12549 non-null  object 
 10  Culture                  10821 non-null  object 
 11  Period                   0 non-null      float64
 12  Dynasty                  0 non-null      float64
 13  Reign                    0 non-null      float64
 14  Portfolio             

~ 🕊 Drop empty columns and delete any duplicate copy rows that may exist

In [5]:
df = df.dropna(axis=1, how='all')

In [6]:
df_no_duplicates = df.drop_duplicates()

Recheck columns removed

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12549 entries, 0 to 12548
Data columns (total 44 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Object Number            12549 non-null  object 
 1   Is Highlight             12549 non-null  bool   
 2   Is Timeline Work         12549 non-null  bool   
 3   Is Public Domain         12549 non-null  bool   
 4   Object ID                12549 non-null  int64  
 5   Gallery Number           8559 non-null   object 
 6   Department               12549 non-null  object 
 7   AccessionYear            12283 non-null  float64
 8   Object Name              12549 non-null  object 
 9   Title                    12549 non-null  object 
 10  Culture                  10821 non-null  object 
 11  Constituent ID           7095 non-null   float64
 12  Artist Role              7095 non-null   object 
 13  Artist Prefix            7095 non-null   object 
 14  Artist Display Name   

Add "Unknown" to missing values in "Artist Gender" column

In [8]:
df['Artist Gender'].fillna('Unknown', inplace=True)

<ipython-input-8-6d3a23f54aae>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Artist Gender'].fillna('Unknown', inplace=True)


~ Check for 12549 entries in Gender (though asking for relevance to project might be useful to the end of simplifying)




In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12549 entries, 0 to 12548
Data columns (total 44 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Object Number            12549 non-null  object 
 1   Is Highlight             12549 non-null  bool   
 2   Is Timeline Work         12549 non-null  bool   
 3   Is Public Domain         12549 non-null  bool   
 4   Object ID                12549 non-null  int64  
 5   Gallery Number           8559 non-null   object 
 6   Department               12549 non-null  object 
 7   AccessionYear            12283 non-null  float64
 8   Object Name              12549 non-null  object 
 9   Title                    12549 non-null  object 
 10  Culture                  10821 non-null  object 
 11  Constituent ID           7095 non-null   float64
 12  Artist Role              7095 non-null   object 
 13  Artist Prefix            7095 non-null   object 
 14  Artist Display Name   

Review the Data Types re: Uniformity

In [39]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12549 entries, 0 to 12548
Data columns (total 40 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Object ID            12549 non-null  int64  
 1   Accession Year       12549 non-null  int64  
 2   Artist Alpha Sort    12549 non-null  object 
 3   Artist Begin Date    5844 non-null   float64
 4   Artist Display Bio   12549 non-null  object 
 5   Artist Display Name  12549 non-null  object 
 6   Artist End Date      5742 non-null   float64
 7   Artist Gender        12549 non-null  object 
 8   Artist Nationality   12549 non-null  object 
 9   Artist Prefix        12549 non-null  object 
 10  Artist Role          12549 non-null  object 
 11  Artist Suffix        12549 non-null  object 
 12  Artist ULAN URL      12549 non-null  object 
 13  Artist Wikidata URL  12549 non-null  object 
 14  City                 12549 non-null  object 
 15  Constituent ID       7095 non-null  

Change Heading to "Accession Year"

In [11]:
df.rename(columns={'AccessionYear': 'Accession Year'}, inplace=True)
df.head()

Object Number  Is Highlight  Is Timeline Work  Is Public Domain  Object ID  \
0    1979.486.1         False             False             False          1   
1    1980.264.5         False             False             False          2   
2      67.265.9         False             False             False          3   
3     67.265.10         False             False             False          4   
4     67.265.11         False             False             False          5   

  Gallery Number         Department  Accession Year Object Name  \
0            NaN  The American Wing          1979.0        Coin   
1            NaN  The American Wing          1980.0        Coin   
2            NaN  The American Wing          1967.0        Coin   
3            NaN  The American Wing          1967.0        Coin   
4            NaN  The American Wing          1967.0        Coin   

                          Title  ... Country  Region Classification  \
0  One-dollar Liberty Head Coin  ...     NaN     NaN            NaN   
1  Ten-dollar Liberty Head Coin  ...     NaN     NaN            NaN   
2    Two-and-a-Half Dollar Coin  ...     NaN     NaN            NaN   
3    Two-and-a-Half Dollar Coin  ...     NaN     NaN            NaN   
4    Two-and-a-Half Dollar Coin  ...     NaN     NaN            NaN   

  Rights and Reproduction                                     Link Resource  \
0                     NaN  http://www.metmuseum.org/art/collection/search/1   
1                     NaN  http://www.metmuseum.org/art/collection/search/2   
2                     NaN  http://www.metmuseum.org/art/collection/search/3   
3                     NaN  http://www.metmuseum.org/art/collection/search/4   
4                     NaN  http://www.metmuseum.org/art/collection/search/5   

  Object Wikidata URL                                Repository Tags  \
0                 NaN  Metropolitan Museum of Art, New York, NY  NaN   
1                 NaN  Metropolitan Museum of Art, New York, NY  NaN   
2                 NaN  Metropolitan Museum of Art, New York, NY  NaN   
3                 NaN  Metropolitan Museum of Art, New York, NY  NaN   
4                 NaN  Metropolitan Museum of Art, New York, NY  NaN   

  Tags AAT URL Tags Wikidata URL  
0          NaN               NaN  
1          NaN               NaN  
2          NaN               NaN  
3          NaN               NaN  
4          NaN               NaN  

[5 rows x 44 columns]

Address Accession Year Dates

In [12]:
df

Object Number  Is Highlight  Is Timeline Work  Is Public Domain  \
0        1979.486.1         False             False             False   
1        1980.264.5         False             False             False   
2          67.265.9         False             False             False   
3         67.265.10         False             False             False   
4         67.265.11         False             False             False   
...             ...           ...               ...               ...   
12544      66.165.9         False             False             False   
12545       36.89.1         False             False              True   
12546       36.89.2         False             False              True   
12547     1985.40.1         False             False             False   
12548      66.165.4         False             False             False   

       Object ID Gallery Number         Department  Accession Year  \
0              1            NaN  The American Wing          1979.0   
1              2            NaN  The American Wing          1980.0   
2              3            NaN  The American Wing          1967.0   
3              4            NaN  The American Wing          1967.0   
4              5            NaN  The American Wing          1967.0   
...          ...            ...                ...             ...   
12544      13714            NaN  The American Wing          1966.0   
12545      13715            NaN  The American Wing          1936.0   
12546      13716            NaN  The American Wing          1936.0   
12547      13717            NaN  The American Wing          1985.0   
12548      13718            NaN  The American Wing          1966.0   

        Object Name                         Title  ...        Country  Region  \
0              Coin  One-dollar Liberty Head Coin  ...            NaN     NaN   
1              Coin  Ten-dollar Liberty Head Coin  ...            NaN     NaN   
2              Coin    Two-and-a-Half Dollar Coin  ...            NaN     NaN   
3              Coin    Two-and-a-Half Dollar Coin  ...            NaN     NaN   
4              Coin    Two-and-a-Half Dollar Coin  ...            NaN     NaN   
...             ...                           ...  ...            ...     ...   
12544  Handkerchief                  Handkerchief  ...  United States     NaN   
12545         Piece                         Piece  ...         France     NaN   
12546         Piece                         Piece  ...         France     NaN   
12547        Length                        Length  ...  United States     NaN   
12548        Length                        Length  ...            NaN     NaN   

      Classification Rights and Reproduction  \
0                NaN                     NaN   
1                NaN                     NaN   
2                NaN                     NaN   
3                NaN                     NaN   
4                NaN                     NaN   
...              ...                     ...   
12544            NaN                     NaN   
12545            NaN                     NaN   
12546            NaN                     NaN   
12547            NaN                     NaN   
12548            NaN                     NaN   

                                           Link Resource Object Wikidata URL  \
0       http://www.metmuseum.org/art/collection/search/1                 NaN   
1       http://www.metmuseum.org/art/collection/search/2                 NaN   
2       http://www.metmuseum.org/art/collection/search/3                 NaN   
3       http://www.metmuseum.org/art/collection/search/4                 NaN   
4       http://www.metmuseum.org/art/collection/search/5                 NaN   
...                                                  ...                 ...   
12544  http://www.metmuseum.org/art/collection/search...                 NaN   
12545  http://www.metmuseum.org/art/collection/search...                 NaN   
12546  http://www.

In [13]:
df['Accession Year'] = df['Accession Year'].fillna(-1).astype(int)


Check if changed to int

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12549 entries, 0 to 12548
Data columns (total 44 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Object Number            12549 non-null  object 
 1   Is Highlight             12549 non-null  bool   
 2   Is Timeline Work         12549 non-null  bool   
 3   Is Public Domain         12549 non-null  bool   
 4   Object ID                12549 non-null  int64  
 5   Gallery Number           8559 non-null   object 
 6   Department               12549 non-null  object 
 7   Accession Year           12549 non-null  int64  
 8   Object Name              12549 non-null  object 
 9   Title                    12549 non-null  object 
 10  Culture                  10821 non-null  object 
 11  Constituent ID           7095 non-null   float64
 12  Artist Role              7095 non-null   object 
 13  Artist Prefix            7095 non-null   object 
 14  Artist Display Name   

~ 👀 Check Number of Missing Values

In [15]:
df.isna().sum()

Object Number                  0
Is Highlight                   0
Is Timeline Work               0
Is Public Domain               0
Object ID                      0
Gallery Number              3990
Department                     0
Accession Year                 0
Object Name                    0
Title                          0
Culture                     1728
Constituent ID              5454
Artist Role                 5454
Artist Prefix               5454
Artist Display Name         5454
Artist Display Bio          5454
Artist Suffix               5454
Artist Alpha Sort           5454
Artist Nationality          5455
Artist Begin Date           5455
Artist End Date             5455
Artist Gender                  0
Artist ULAN URL             7545
Artist Wikidata URL         7764
Object Date                  294
Object Begin Date              1
Object End Date                1
Medium                         2
Dimensions                     4
Credit Line                    4
Geography Type              4321
City                        7883
State                      11830
County                     11829
Country                     4323
Region                     12181
Classification             12545
Rights and Reproduction    12476
Link Resource                  1
Object Wikidata URL        10654
Repository                     1
Tags                        7098
Tags AAT URL                7098
Tags Wikidata URL           7098
dtype: int64

~ Drop Columns with too many missing entries to be useful or that clearly contain incorrect information. In this case, entires in County and Region appear to be repeated for both columns and these are clearly connected - possibly an entry error  

In [16]:
to_drop = ['County',
'Region',
'Classification',
'Rights and Reproduction']
df.drop(to_drop, inplace = True, axis = 1)

~ Check what Columns I have left

In [17]:
print(df.columns)

Index(['Object Number', 'Is Highlight', 'Is Timeline Work', 'Is Public Domain',
       'Object ID', 'Gallery Number', 'Department', 'Accession Year',
       'Object Name', 'Title', 'Culture', 'Constituent ID', 'Artist Role',
       'Artist Prefix', 'Artist Display Name', 'Artist Display Bio',
       'Artist Suffix', 'Artist Alpha Sort', 'Artist Nationality',
       'Artist Begin Date', 'Artist End Date', 'Artist Gender',
       'Artist ULAN URL', 'Artist Wikidata URL', 'Object Date',
       'Object Begin Date', 'Object End Date', 'Medium', 'Dimensions',
       'Credit Line', 'Geography Type', 'City', 'State', 'Country',
       'Link Resource', 'Object Wikidata URL', 'Repository', 'Tags',
       'Tags AAT URL', 'Tags Wikidata URL'],
      dtype='object')


In [18]:
df.head()

Object Number  Is Highlight  Is Timeline Work  Is Public Domain  Object ID  \
0    1979.486.1         False             False             False          1   
1    1980.264.5         False             False             False          2   
2      67.265.9         False             False             False          3   
3     67.265.10         False             False             False          4   
4     67.265.11         False             False             False          5   

  Gallery Number         Department  Accession Year Object Name  \
0            NaN  The American Wing            1979        Coin   
1            NaN  The American Wing            1980        Coin   
2            NaN  The American Wing            1967        Coin   
3            NaN  The American Wing            1967        Coin   
4            NaN  The American Wing            1967        Coin   

                          Title  ... Geography Type  City State Country  \
0  One-dollar Liberty Head Coin  ...            NaN   NaN   NaN     NaN   
1  Ten-dollar Liberty Head Coin  ...            NaN   NaN   NaN     NaN   
2    Two-and-a-Half Dollar Coin  ...            NaN   NaN   NaN     NaN   
3    Two-and-a-Half Dollar Coin  ...            NaN   NaN   NaN     NaN   
4    Two-and-a-Half Dollar Coin  ...            NaN   NaN   NaN     NaN   

                                      Link Resource Object Wikidata URL  \
0  http://www.metmuseum.org/art/collection/search/1                 NaN   
1  http://www.metmuseum.org/art/collection/search/2                 NaN   
2  http://www.metmuseum.org/art/collection/search/3                 NaN   
3  http://www.metmuseum.org/art/collection/search/4                 NaN   
4  http://www.metmuseum.org/art/collection/search/5                 NaN   

                                 Repository Tags Tags AAT URL  \
0  Metropolitan Museum of Art, New York, NY  NaN          NaN   
1  Metropolitan Museum of Art, New York, NY  NaN          NaN   
2  Metropolitan Museum of Art, New York, NY  NaN          NaN   
3  Metropolitan Museum of Art, New York, NY  NaN          NaN   
4  Metropolitan Museum of Art, New York, NY  NaN          NaN   

  Tags Wikidata URL  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4               NaN  

[5 rows x 40 columns]

In [19]:
df.isna().sum()

Object Number              0
Is Highlight               0
Is Timeline Work           0
Is Public Domain           0
Object ID                  0
Gallery Number          3990
Department                 0
Accession Year             0
Object Name                0
Title                      0
Culture                 1728
Constituent ID          5454
Artist Role             5454
Artist Prefix           5454
Artist Display Name     5454
Artist Display Bio      5454
Artist Suffix           5454
Artist Alpha Sort       5454
Artist Nationality      5455
Artist Begin Date       5455
Artist End Date         5455
Artist Gender              0
Artist ULAN URL         7545
Artist Wikidata URL     7764
Object Date              294
Object Begin Date          1
Object End Date            1
Medium                     2
Dimensions                 4
Credit Line                4
Geography Type          4321
City                    7883
State                  11830
Country                 4323
Link Resource              1
Object Wikidata URL    10654
Repository                 1
Tags                    7098
Tags AAT URL            7098
Tags Wikidata URL       7098
dtype: int64

🍎 Check on Data Overview


In [20]:
df

Object Number  Is Highlight  Is Timeline Work  Is Public Domain  \
0        1979.486.1         False             False             False   
1        1980.264.5         False             False             False   
2          67.265.9         False             False             False   
3         67.265.10         False             False             False   
4         67.265.11         False             False             False   
...             ...           ...               ...               ...   
12544      66.165.9         False             False             False   
12545       36.89.1         False             False              True   
12546       36.89.2         False             False              True   
12547     1985.40.1         False             False             False   
12548      66.165.4         False             False             False   

       Object ID Gallery Number         Department  Accession Year  \
0              1            NaN  The American Wing            1979   
1              2            NaN  The American Wing            1980   
2              3            NaN  The American Wing            1967   
3              4            NaN  The American Wing            1967   
4              5            NaN  The American Wing            1967   
...          ...            ...                ...             ...   
12544      13714            NaN  The American Wing            1966   
12545      13715            NaN  The American Wing            1936   
12546      13716            NaN  The American Wing            1936   
12547      13717            NaN  The American Wing            1985   
12548      13718            NaN  The American Wing            1966   

        Object Name                         Title  ...   Geography Type  \
0              Coin  One-dollar Liberty Head Coin  ...              NaN   
1              Coin  Ten-dollar Liberty Head Coin  ...              NaN   
2              Coin    Two-and-a-Half Dollar Coin  ...              NaN   
3              Coin    Two-and-a-Half Dollar Coin  ...              NaN   
4              Coin    Two-and-a-Half Dollar Coin  ...              NaN   
...             ...                           ...  ...              ...   
12544  Handkerchief                  Handkerchief  ...  (none assigned)   
12545         Piece                         Piece  ...          Made in   
12546         Piece                         Piece  ...          Made in   
12547        Length                        Length  ...          Made in   
12548        Length                        Length  ...              NaN   

              City State        Country  \
0              NaN   NaN            NaN   
1              NaN   NaN            NaN   
2              NaN   NaN            NaN   
3              NaN   NaN            NaN   
4              NaN   NaN            NaN   
...            ...   ...            ...   
12544  New Lebanon   NaN  United States   
12545          NaN   NaN         France   
12546          NaN   NaN         France   
12547  New Lebanon   NaN  United States   
12548          NaN   NaN            NaN   

                                           Link Resource Object Wikidata URL  \
0       http://www.metmuseum.org/art/collection/search/1                 NaN   
1       http://www.metmuseum.org/art/collection/search/2                 NaN   
2       http://www.metmuseum.org/art/collection/search/3                 NaN   
3       http://www.metmuseum.org/art/collection/search/4                 NaN   
4       http://www.metmuseum.org/art/collection/search/5                 NaN   
...                                                  ...                 ...   
12544  http://www.metmuseum.org/art/collection/search...                 NaN   
12545  http://www.metmuseum.org/art/collection/search...                 NaN   
12546  http://www.metmuseum.org/art/collection/search...                 NaN   
12547  http://www.metmuseum.org/art/collection/search...              

Sort Columns ABC

In [21]:
sorted_columns = sorted(df.columns)
df = df.reindex(columns=sorted_columns)
df.head()

Accession Year       Artist Alpha Sort Artist Begin Date  \
0            1979  Longacre, James Barton              1794   
1            1980     Gobrecht, Christian              1785   
2            1967                     NaN               NaN   
3            1967                     NaN               NaN   
4            1967                     NaN               NaN   

                                  Artist Display Bio    Artist Display Name  \
0  American, Delaware County, Pennsylvania 1794–1...  James Barton Longacre   
1                                          1785–1844     Christian Gobrecht   
2                                                NaN                    NaN   
3                                                NaN                    NaN   
4                                                NaN                    NaN   

  Artist End Date Artist Gender Artist Nationality Artist Prefix Artist Role  \
0            1869       Unknown           American                     Maker   
1            1844       Unknown                                        Maker   
2             NaN       Unknown                NaN           NaN         NaN   
3             NaN       Unknown                NaN           NaN         NaN   
4             NaN       Unknown                NaN           NaN         NaN   

   ... Object ID Object Name Object Number Object Wikidata URL  \
0  ...         1        Coin    1979.486.1                 NaN   
1  ...         2        Coin    1980.264.5                 NaN   
2  ...         3        Coin      67.265.9                 NaN   
3  ...         4        Coin     67.265.10                 NaN   
4  ...         5        Coin     67.265.11                 NaN   

                                 Repository State Tags Tags AAT URL  \
0  Metropolitan Museum of Art, New York, NY   NaN  NaN          NaN   
1  Metropolitan Museum of Art, New York, NY   NaN  NaN          NaN   
2  Metropolitan Museum of Art, New York, NY   NaN  NaN          NaN   
3  Metropolitan Museum of Art, New York, NY   NaN  NaN          NaN   
4  Metropolitan Museum of Art, New York, NY   NaN  NaN          NaN   

  Tags Wikidata URL                         Title  
0               NaN  One-dollar Liberty Head Coin  
1               NaN  Ten-dollar Liberty Head Coin  
2               NaN    Two-and-a-Half Dollar Coin  
3               NaN    Two-and-a-Half Dollar Coin  
4               NaN    Two-and-a-Half Dollar Coin  

[5 rows x 40 columns]

Move Object ID to First Column

In [22]:
columns = df.columns.tolist()

In [23]:
cols = list(df.columns)

In [24]:
cols.insert(0, cols.pop(cols.index('Object ID')))

In [25]:
df = df[cols]

In [26]:
df

Object ID  Accession Year  \
0              1            1979   
1              2            1980   
2              3            1967   
3              4            1967   
4              5            1967   
...          ...             ...   
12544      13714            1966   
12545      13715            1936   
12546      13716            1936   
12547      13717            1985   
12548      13718            1966   

                                       Artist Alpha Sort Artist Begin Date  \
0                                 Longacre, James Barton              1794   
1                                    Gobrecht, Christian              1785   
2                                                    NaN               NaN   
3                                                    NaN               NaN   
4                                                    NaN               NaN   
...                                                  ...               ...   
12544  United Society of Believers in Christ’s Second...              1750   
12545                             Huet, Jean-Baptiste, I              1745   
12546                             Huet, Jean-Baptiste, I              1745   
12547  United Society of Believers in Christ’s Second...              1750   
12548  Jacobs, Clarissa|North Family |United Society ...               NaN   

                                      Artist Display Bio  \
0      American, Delaware County, Pennsylvania 1794–1...   
1                                              1785–1844   
2                                                    NaN   
3                                                    NaN   
4                                                    NaN   
...                                                  ...   
12544                  American, active ca. 1750–present   
12545                      French, Paris 1745–1811 Paris   
12546                      French, Paris 1745–1811 Paris   
12547                  American, active ca. 1750–present   
12548               | |American, active ca. 1750–present   

                                     Artist Display Name Artist End Date  \
0                                  James Barton Longacre            1869   
1                                     Christian Gobrecht            1844   
2                                                    NaN             NaN   
3                                                    NaN             NaN   
4                                                    NaN             NaN   
...                                                  ...             ...   
12544  United Society of Believers in Christ’s Second...            9999   
12545                               Jean-Baptiste Huet I            1811   
12546                               Jean-Baptiste Huet I            1811   
12547  United Society of Believers in Christ’s Second...            9999   
12548  Clarissa Jacobs|North Family|United Society of...             NaN   

      Artist Gender Artist Nationality Artist Prefix  ... Object End Date  \
0           Unknown           American                ...          1853.0   
1           Unknown                                   ...          1901.0   
2           Unknown                NaN           NaN  ...          1927.0   
3           Unknown                NaN           NaN  ...          1927.0   
4           Unknown                NaN           NaN  ...          1927.0   
...             ...                ...           ...  ...             ...   
12544       Unknown                                   ...          1900.0   
12545       Unknown             French                ...          1788.0   
12546       Unknown             French                ...          1788.0   
12547       Unknown                                   ...          1900.0   
12548       Unknown                NaN          | |   ...             NaN   

        Object Name Object Number Object Wikidata URL  \
0              Coin  

(Above shows that ID numbers for the museum have not all been accounted for, assuming they started at 1 and went up sequentially. This may be useful in assessing how complete their inventory process has been thus far and/or examining loss of pieces?)

In [27]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12549 entries, 0 to 12548
Data columns (total 40 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Object ID            12549 non-null  int64  
 1   Accession Year       12549 non-null  int64  
 2   Artist Alpha Sort    7095 non-null   object 
 3   Artist Begin Date    7094 non-null   object 
 4   Artist Display Bio   7095 non-null   object 
 5   Artist Display Name  7095 non-null   object 
 6   Artist End Date      7094 non-null   object 
 7   Artist Gender        12549 non-null  object 
 8   Artist Nationality   7094 non-null   object 
 9   Artist Prefix        7095 non-null   object 
 10  Artist Role          7095 non-null   object 
 11  Artist Suffix        7095 non-null   object 
 12  Artist ULAN URL      5004 non-null   object 
 13  Artist Wikidata URL  4785 non-null   object 
 14  City                 4666 non-null   object 
 15  Constituent ID       7095 non-null  

Clean Date Formats

In [40]:
date_columns = ['Artist Begin Date', 'Artist End Date', 'Object End Date','Object Date']

for col in date_columns:
    # Convert to datetime
    df[col] = pd.to_datetime(df[col], errors='coerce')

    # Extract year and replace the original column
    df[col] = df[col].dt.year

    # Convert to integer while handling NaN values
    df[col] = df[col].astype('Int64')



In [41]:
df.head()

Object ID  Accession Year       Artist Alpha Sort  Artist Begin Date  \
0          1            1979  Longacre, James Barton               1970   
1          2            1980     Gobrecht, Christian               1970   
2          3            1967                 Unknown               <NA>   
3          4            1967                 Unknown               <NA>   
4          5            1967                 Unknown               <NA>   

                                  Artist Display Bio    Artist Display Name  \
0  American, Delaware County, Pennsylvania 1794–1...  James Barton Longacre   
1                                          1785–1844     Christian Gobrecht   
2                                            Unknown                Unknown   
3                                            Unknown                Unknown   
4                                            Unknown                Unknown   

   Artist End Date Artist Gender Artist Nationality Artist Prefix  ...  \
0             1970       Unknown           American       Unknown  ...   
1             1970       Unknown            Unknown       Unknown  ...   
2             <NA>       Unknown            Unknown       Unknown  ...   
3             <NA>       Unknown            Unknown       Unknown  ...   
4             <NA>       Unknown            Unknown       Unknown  ...   

  Object End Date Object Name Object Number Object Wikidata URL  \
0            1970        Coin    1979.486.1             Unknown   
1            1970        Coin    1980.264.5             Unknown   
2            1970        Coin      67.265.9             Unknown   
3            1970        Coin     67.265.10             Unknown   
4            1970        Coin     67.265.11             Unknown   

                                 Repository    State     Tags Tags AAT URL  \
0  Metropolitan Museum of Art, New York, NY  Unknown  Unknown      Unknown   
1  Metropolitan Museum of Art, New York, NY  Unknown  Unknown      Unknown   
2  Metropolitan Museum of Art, New York, NY  Unknown  Unknown      Unknown   
3  Metropolitan Museum of Art, New York, NY  Unknown  Unknown      Unknown   
4  Metropolitan Museum of Art, New York, NY  Unknown  Unknown      Unknown   

  Tags Wikidata URL                         Title  
0           Unknown  One-dollar Liberty Head Coin  
1           Unknown  Ten-dollar Liberty Head Coin  
2           Unknown    Two-and-a-Half Dollar Coin  
3           Unknown    Two-and-a-Half Dollar Coin  
4           Unknown    Two-and-a-Half Dollar Coin  

[5 rows x 40 columns]

Strip Whitespace

In [42]:
df = df.apply(lambda x: x.str.strip() if x.dtype == "object" else x)


In [43]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12549 entries, 0 to 12548
Data columns (total 40 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Object ID            12549 non-null  int64  
 1   Accession Year       12549 non-null  int64  
 2   Artist Alpha Sort    12549 non-null  object 
 3   Artist Begin Date    5844 non-null   Int64  
 4   Artist Display Bio   12549 non-null  object 
 5   Artist Display Name  12549 non-null  object 
 6   Artist End Date      5742 non-null   Int64  
 7   Artist Gender        12549 non-null  object 
 8   Artist Nationality   12549 non-null  object 
 9   Artist Prefix        12549 non-null  object 
 10  Artist Role          12549 non-null  object 
 11  Artist Suffix        12549 non-null  object 
 12  Artist ULAN URL      12549 non-null  object 
 13  Artist Wikidata URL  12549 non-null  object 
 14  City                 12549 non-null  object 
 15  Constituent ID       7095 non-null  

In [32]:
# Function to remove blank space entries
def remove_blank_spaces(x):
    if isinstance(x, str):
        return x.strip() if x.strip() else pd.NA
    return x

# Apply the function to all elements in the DataFrame
df = df.map(remove_blank_spaces)

# Remove any rows that are now entirely NA
df = df.dropna(how='all')

# Reset the index if needed
df = df.reset_index(drop=True)

In [33]:
object_columns = df.select_dtypes(include=['object']).columns

for column in object_columns:
    df[column] = df[column].fillna('Unknown')

In [34]:
df

Object ID  Accession Year  \
0              1            1979   
1              2            1980   
2              3            1967   
3              4            1967   
4              5            1967   
...          ...             ...   
12544      13714            1966   
12545      13715            1936   
12546      13716            1936   
12547      13717            1985   
12548      13718            1966   

                                       Artist Alpha Sort  Artist Begin Date  \
0                                 Longacre, James Barton             1794.0   
1                                    Gobrecht, Christian             1785.0   
2                                                Unknown                NaN   
3                                                Unknown                NaN   
4                                                Unknown                NaN   
...                                                  ...                ...   
12544  United Society of Believers in Christ’s Second...             1750.0   
12545                             Huet, Jean-Baptiste, I             1745.0   
12546                             Huet, Jean-Baptiste, I             1745.0   
12547  United Society of Believers in Christ’s Second...             1750.0   
12548  Jacobs, Clarissa|North Family |United Society ...                NaN   

                                      Artist Display Bio  \
0      American, Delaware County, Pennsylvania 1794–1...   
1                                              1785–1844   
2                                                Unknown   
3                                                Unknown   
4                                                Unknown   
...                                                  ...   
12544                  American, active ca. 1750–present   
12545                      French, Paris 1745–1811 Paris   
12546                      French, Paris 1745–1811 Paris   
12547                  American, active ca. 1750–present   
12548               | |American, active ca. 1750–present   

                                     Artist Display Name  Artist End Date  \
0                                  James Barton Longacre           1869.0   
1                                     Christian Gobrecht           1844.0   
2                                                Unknown              NaN   
3                                                Unknown              NaN   
4                                                Unknown              NaN   
...                                                  ...              ...   
12544  United Society of Believers in Christ’s Second...              NaN   
12545                               Jean-Baptiste Huet I           1811.0   
12546                               Jean-Baptiste Huet I           1811.0   
12547  United Society of Believers in Christ’s Second...              NaN   
12548  Clarissa Jacobs|North Family|United Society of...              NaN   

      Artist Gender Artist Nationality Artist Prefix  ... Object End Date  \
0           Unknown           American       Unknown  ...          1970.0   
1           Unknown            Unknown       Unknown  ...          1970.0   
2           Unknown            Unknown       Unknown  ...          1970.0   
3           Unknown            Unknown       Unknown  ...          1970.0   
4           Unknown            Unknown       Unknown  ...          1970.0   
...             ...                ...           ...  ...             ...   
12544       Unknown            Unknown       Unknown  ...          1970.0   
12545       Unknown             French       Unknown  ...          1970.0   
12546       Unknown             French       Unknown  ...          1970.0   
12547       Unknown            Unknown       Unknown  ...          1970.0   
12548       Unknown            Unknown           | |  ...             NaN   

        Object Name Object Number Object Wikidata URL 

Replace NaN with "Unknown"

In [35]:
df

Object ID  Accession Year  \
0              1            1979   
1              2            1980   
2              3            1967   
3              4            1967   
4              5            1967   
...          ...             ...   
12544      13714            1966   
12545      13715            1936   
12546      13716            1936   
12547      13717            1985   
12548      13718            1966   

                                       Artist Alpha Sort  Artist Begin Date  \
0                                 Longacre, James Barton             1794.0   
1                                    Gobrecht, Christian             1785.0   
2                                                Unknown                NaN   
3                                                Unknown                NaN   
4                                                Unknown                NaN   
...                                                  ...                ...   
12544  United Society of Believers in Christ’s Second...             1750.0   
12545                             Huet, Jean-Baptiste, I             1745.0   
12546                             Huet, Jean-Baptiste, I             1745.0   
12547  United Society of Believers in Christ’s Second...             1750.0   
12548  Jacobs, Clarissa|North Family |United Society ...                NaN   

                                      Artist Display Bio  \
0      American, Delaware County, Pennsylvania 1794–1...   
1                                              1785–1844   
2                                                Unknown   
3                                                Unknown   
4                                                Unknown   
...                                                  ...   
12544                  American, active ca. 1750–present   
12545                      French, Paris 1745–1811 Paris   
12546                      French, Paris 1745–1811 Paris   
12547                  American, active ca. 1750–present   
12548               | |American, active ca. 1750–present   

                                     Artist Display Name  Artist End Date  \
0                                  James Barton Longacre           1869.0   
1                                     Christian Gobrecht           1844.0   
2                                                Unknown              NaN   
3                                                Unknown              NaN   
4                                                Unknown              NaN   
...                                                  ...              ...   
12544  United Society of Believers in Christ’s Second...              NaN   
12545                               Jean-Baptiste Huet I           1811.0   
12546                               Jean-Baptiste Huet I           1811.0   
12547  United Society of Believers in Christ’s Second...              NaN   
12548  Clarissa Jacobs|North Family|United Society of...              NaN   

      Artist Gender Artist Nationality Artist Prefix  ... Object End Date  \
0           Unknown           American       Unknown  ...          1970.0   
1           Unknown            Unknown       Unknown  ...          1970.0   
2           Unknown            Unknown       Unknown  ...          1970.0   
3           Unknown            Unknown       Unknown  ...          1970.0   
4           Unknown            Unknown       Unknown  ...          1970.0   
...             ...                ...           ...  ...             ...   
12544       Unknown            Unknown       Unknown  ...          1970.0   
12545       Unknown             French       Unknown  ...          1970.0   
12546       Unknown             French       Unknown  ...          1970.0   
12547       Unknown            Unknown       Unknown  ...          1970.0   
12548       Unknown            Unknown           | |  ...             NaN   

        Object Name Object Number Object Wikidata URL 

In [36]:
df.dtypes

Object ID                int64
Accession Year           int64
Artist Alpha Sort       object
Artist Begin Date      float64
Artist Display Bio      object
Artist Display Name     object
Artist End Date        float64
Artist Gender           object
Artist Nationality      object
Artist Prefix           object
Artist Role             object
Artist Suffix           object
Artist ULAN URL         object
Artist Wikidata URL     object
City                    object
Constituent ID         float64
Country                 object
Credit Line             object
Culture                 object
Department              object
Dimensions              object
Gallery Number          object
Geography Type          object
Is Highlight              bool
Is Public Domain          bool
Is Timeline Work          bool
Link Resource           object
Medium                  object
Object Begin Date      float64
Object Date            float64
Object End Date        float64
Object Name             object
Object Number           object
Object Wikidata URL     object
Repository              object
State                   object
Tags                    object
Tags AAT URL            object
Tags Wikidata URL       object
Title                   object
dtype: object

In [37]:
df.describe()

Object ID  Accession Year  Artist Begin Date  Artist End Date  \
count  12549.000000    12549.000000        5844.000000      5742.000000   
mean    6772.089808     1901.652642        1816.916324      1877.381574   
std     3939.600059      281.239958          46.873479        51.623705   
min        1.000000       -1.000000        1678.000000      1681.000000   
25%     3357.000000     1922.000000        1785.000000      1844.000000   
50%     6678.000000     1946.000000        1826.000000      1888.000000   
75%    10154.000000     1965.000000        1856.000000      1922.000000   
max    13718.000000     2021.000000        1918.000000      2063.000000   

       Constituent ID  Object Begin Date  Object Date  Object End Date  
count    7.095000e+03       12548.000000  1867.000000          12548.0  
mean     1.682482e+17        1822.050765  1862.569363           1970.0  
std      1.415345e+19          63.418491    43.815198              0.0  
min      1.070000e+02           0.000000  1679.000000           1970.0  
25%      1.480000e+02        1787.000000  1840.500000           1970.0  
50%      2.990000e+02        1828.000000  1872.000000           1970.0  
75%      1.625300e+04        1870.000000  1893.000000           1970.0  
max      1.192170e+21        1976.000000  1976.000000           1970.0

In [38]:
df.isna().sum()

Object ID                  0
Accession Year             0
Artist Alpha Sort          0
Artist Begin Date       6705
Artist Display Bio         0
Artist Display Name        0
Artist End Date         6807
Artist Gender              0
Artist Nationality         0
Artist Prefix              0
Artist Role                0
Artist Suffix              0
Artist ULAN URL            0
Artist Wikidata URL        0
City                       0
Constituent ID          5454
Country                    0
Credit Line                0
Culture                    0
Department                 0
Dimensions                 0
Gallery Number             0
Geography Type             0
Is Highlight               0
Is Public Domain           0
Is Timeline Work           0
Link Resource              0
Medium                     0
Object Begin Date          1
Object Date            10682
Object End Date            1
Object Name                0
Object Number              0
Object Wikidata URL        0
Repository                 0
State                      0
Tags                       0
Tags AAT URL               0
Tags Wikidata URL          0
Title                      0
dtype: int64